File: Dynamic Programming/Number of Sub-arrays With Odd Sum.go
func numOfSubarrays(arr []int) int {
    odd_sum := 0
    even_sum := 0
    res := 0
    prefix := 0

    for _, val := range arr {
        prefix += val
        if (prefix & 1 == 1) {
            odd_sum += 1
            res += even_sum + 1
        } else {
            even_sum += 1
            res += odd_sum
        }
    }

    return res % (1e9 + 7)
}

File: Dynamic Programming/Number of Sub-arrays With Odd Sum.rs
impl Solution {
    pub fn num_of_subarrays(arr: Vec<i32>) -> i32 {
        let mut res = 0_i64;
        let mut sum = 0;
        let mut prev_odd = 0;
        let mut prev_even = 0;

        for i in 0..arr.len() {
            sum += arr[i];
            if sum & 1 == 1 {
                prev_odd += 1;
                res += prev_even + 1;
            } else {
                prev_even += 1;
                res += prev_odd;
            }
        }
        (res % 1000000007) as i32
    }
}